README.md
# slidy_calendar
[![Pub Version](https://img.shields.io/pub/v/flutter_calendar_carousel.svg?style=flat-square)](https://pub.dartlang.org/packages/flutter_calendar_carousel)
[![Build Status](https://travis-ci.com/dooboolab/flutter_calendar_carousel.svg?branch=master)](https://travis-ci.com/dooboolab/flutter_calendar_carousel)
![License](https://img.shields.io/badge/license-MIT-blue.svg)
[![Coverage Status](https://coveralls.io/repos/github/dooboolab/flutter_calendar_carousel/badge.svg?branch=master)](https://coveralls.io/github/dooboolab/flutter_calendar_carousel?branch=master)

Calendar widget for flutter that is swipeable horizontally. This widget can help you build your own calendar widget highly customizable. Now you can even add your icon for each event.

## New Feature
+ Check out great feature `customDayBuilder` work done by [maxgmer](https://github.com/maxgmer) :tada:.

#### Rectangular style
![image](https://raw.githubusercontent.com/dooboolab/flutter_calendar_carousel/master/doc/calendar1.gif)

#### Circular style
![image](https://raw.githubusercontent.com/dooboolab/flutter_calendar_carousel/master/doc/calendar2.gif)

#### No border
![image](https://raw.githubusercontent.com/dooboolab/flutter_calendar_carousel/master/doc/calendar3.gif)

#### Marked Dates
![image](https://raw.githubusercontent.com/dooboolab/flutter_calendar_carousel/master/doc/calendar4.gif)

#### Custom Icon Events
![image](https://raw.githubusercontent.com/icemanbsi/flutter_calendar_carousel/master/doc/calendar5.gif)

## Getting Started
For help getting started with Flutter, view our online
[documentation](https://flutter.io/).

## Props
| props                   | types           | defaultValues                                                                                                     |
| :---------------------- | :-------------: | :---------------------------------------------------------------------------------------------------------------: |
| viewPortFraction        | `double`        | 1.0 |
| prevDaysTextStyle       | `TextStyle`     | |
| daysTextStyle           | `TextStyle`     | |
| nextDaysTextStyle       | `TextStyle`     | |
| prevMonthDayBorderColor | `Color`         | Colors.transparent |
| thisMonthDayBorderColor | `Color`         | Colors.transparent |
| nextMonthDayBorderColor | `Color`         | Colors.transparent |
| dayPadding              | `double`        | 2.0 |
| height                  | `double`        | double.infinity |
| width                   | `double`        | double.infinity |
| todayTextStyle          | `TextStyle`     | `fontSize: 14.0, color: Colors.white` |
| dayButtonColor          | `Color`         | Colors.red |
| todayBorderColor        | `Color`         | Colors.red |
| todayButtonColor        | `Colors`        | Colors.red |
| selectedDateTime        | `DateTime`      | |
| selectedDayTextStyle    | `TextStyle`     | `fontSize: 14.0, color: Colors.white` |
| selectedDayBorderColor  | `Color`         | Colors.green |
| selectedDayButtonColor  | `Color`         | Colors.green |
| daysHaveCircularBorder  | `bool`          | |
| onDayPressed            | `Func`          | |
| weekdayTextStyle        | `TextStyle`     | `fontSize: 14.0, color: Colors.deepOrange` |
| iconColor               | `Color`         | Colors.blueAccent |
| headerTextStyle         | `TextStyle`     | `fontSize: 20.0, color: Colors.blue` |
| headerText              | `Text`          | `Text('${DateFormat.yMMM().format(this._dates[1])}'`) |
| weekendTextStyle        | `TextStyle`     | `fontSize: 14.0, color: Colors.pinkAccent` |
| markedDatesMap          | `Events`        | `null` |
| markedDateWidget        | `Color`         | ``` Positioned(child: Container(color: Colors.blueAccent, height: 4.0, width: 4.0), bottom: 4.0, left: 18.0); ``` |
| markedDateShowIcon      | `bool`          | false |
| markedDateIconBorderColor | `Color`       | |
| markedDateIconMaxShown  | `int`           | 2 |
| markedDateIconMargin    | `double`        | 5.0 |
| markedDateIconBuilder   | `MarkedDateIconBuilder<T>` | |
| markedDateIconOffset    | `double`        | 5.0 |
| markedDateCustomShapeBorder | `ShapeBorder` | null |
| markedDateCustomTextStyle | `TextStyle` | null |
| markedDateMoreCustomDecoration | `Decoration` |    |
| markedDateMoreCustomTextStyle | `TextStyle` |   |
| headerMargin            | `EdgetInsets`   | `const EdgeInsets.symmetric(vertical: 16.0)` |
| headerTitleTouchable    | `bool`          | `false` |
| onHeaderTitlePressed    | `Function`      | `() => _selectDateFromPicker()` |
| showHeader              | `bool`          | |
| showHeaderButton        | `bool`          | |
| childAspectRatio        | `double`        | `1.0` |
| weekDayMargin           | `EdgeInsets`    | `const EdgeInsets.only(bottom: 4.0)` |
| weekFormat              | `bool`          | `false` |
| locale                  | `String`        | `en` |
| firstDayOfWeek          | `int`           | `null` |
| onCalendarChanged       | `Function(DateTime)` | |
| minSelectedDate         | `DateTime`      | |
| maxSelectedDate         | `DateTime`      | |
| inactiveDaysTextStyle   | `TextStyle`     | |
| inactiveWeekendTextStyle | `TextStyle`    | |
| weekDayFormat           | `WeekdayFormat` | `short` |
| staticSixWeekFormat     | `bool`          | `false` |
| showOnlyCurrentMonthDate | `bool`          | `false` |
| dayCrossAxisAlignment | `CrossAxisAlignment` | `CrossAxisAlignment.center` |
| dayMainAxisAlignment | `MainAxisAlignment` | `CrossAlignment.center` |
| showIconBehindDayText | `bool` | `false` |
| pageScrollPhysics | `ScrollPhysics` | `ScrollPhysics` |

With ``CalendarCarousel<YourEventClass>`` and ``EventList<YourEventClass>`` you can specifiy a custom Event class.

## Install
Add ```flutter_calendar_carousel``` as a dependency in pubspec.yaml
For help on adding as a dependency, view the [documentation](https://flutter.io/using-packages/).

## Usage
```dart
import 'package:flutter_calendar_carousel/flutter_calendar_carousel.dart' show CalendarCarousel;
Widget widget() {
  return Container(
    margin: EdgeInsets.symmetric(horizontal: 16.0),
    child: CalendarCarousel<Event>(
      onDayPressed: (DateTime date, List<Event> events) {
        this.setState(() => _currentDate = date);
      },
      weekendTextStyle: TextStyle(
        color: Colors.red,
      ),
      thisMonthDayBorderColor: Colors.grey,
//      weekDays: null, /// for pass null when you do not want to render weekDays
//      headerText: Container( /// Example for rendering custom header
//        child: Text('Custom Header'),
//      ),
      customDayBuilder: (   /// you can provide your own build function to make custom day containers
        bool isSelectable,
        int index,
        bool isSelectedDay,
        bool isToday,
        bool isPrevMonthDay,
        TextStyle textStyle,
        bool isNextMonthDay,
        bool isThisMonthDay,
        DateTime day,
      ) {
          /// If you return null, [CalendarCarousel] will build container for current [day] with default function.
          /// This way you can build custom containers for specific days only, leaving rest as default.

          // Example: every 15th of month, we have a flight, we can place an icon in the container like that:
          if (day.day == 15) {
            return Center(
              child: Icon(Icons.local_airport),
            );
          } else {
            return null;
          }
      },
      weekFormat: false,
      markedDatesMap: _markedDateMap,
      height: 420.0,
      selectedDateTime: _currentDate,
      daysHaveCircularBorder: false, /// null for not rendering any border, true for circular border, false for rectangular border
    ),
  );
}
```

### TODO
- [x] Render weekdays.
- [x] Customizable headerWidget.
- [x] Set weekdays visibility.
- [x] Customizable textStyles for days in weekend.
- [x] Marked Dates.
- [x] Multiple Marked Dates.
- [x] Customizable weekend days.
- [x] Week Calendar.
- [x] Carousel Week Calendar.
- [ ] Multiple days selections. 
- [x] Widget test. 

## Help Maintenance
I've been maintaining quite many repos these days and burning out slowly. If you could help me cheer up, buying me a cup of coffee will make my life really happy and get much energy out of it.
<br/>
<a href="https://www.buymeacoffee.com/dooboolab" target="_blank"><img src="https://www.buymeacoffee.com/assets/img/custom_images/purple_img.png" alt="Buy Me A Coffee" style="height: auto !important;width: auto !important;" ></a>
[![Paypal](https://www.paypalobjects.com/webstatic/mktg/Logo/pp-logo-100px.png)](https://paypal.me/dooboolab)


CHANGELOG.md
## [1.5.2]

- Forked by flutter_calendar_carousel (https://github.com/dooboolab/flutter_calendar_carousel)
  
## [1.5.1]

- Bugfix when switching month - "The method 'call' was called on null." [#243](https://github.com/dooboolab/flutter_calendar_carousel/pull/243)

## [1.5.0]

- Add key to widget constructor [#234](https://github.com/dooboolab/flutter_calendar_carousel/pull/234/files)
- Enhance initilizing page numbers [#231](https://github.com/dooboolab/flutter_calendar_carousel/pull/231)

## [1.4.12]

- Handle issue [#207](https://github.com/dooboolab/flutter_calendar_carousel/issues/207), [#209](https://github.com/dooboolab/flutter_calendar_carousel/issues/209)

## [1.4.11]

- Add first day of week offset to week builder [#204](https://github.com/dooboolab/flutter_calendar_carousel/pull/204)

## [1.4.10]

- Fix Calendar displays incorrectly when scrolling horizontally [#193](https://github.com/dooboolab/flutter_calendar_carousel/pull/193)

## [1.4.9]

- Target date for custom header

## [1.4.8]

- Add ability to set `targetDate` on header [#183](https://github.com/dooboolab/flutter_calendar_carousel/pull/183).

## [1.4.7]

- Fix current day showing incorrectly when using `showOnlyCurrentMonthDate` [#181](https://github.com/dooboolab/flutter_calendar_carousel/pull/182).

## [1.4.6]

- Set default `minSelectedDate` and `maxSelectedDate` [#179](https://github.com/dooboolab/flutter_calendar_carousel/pull/179).

## [1.4.4]

- Expose `pageScrollPhysics` for pageView.

## [1.4.2]

- Add option for setting scrollDirection [#166](https://github.com/dooboolab/flutter_calendar_carousel/pull/166)
- Resolve [#123](https://github.com/dooboolab/flutter_calendar_carousel/issues/123) in [#165](https://github.com/dooboolab/flutter_calendar_carousel/pull/165).

## [1.4.1]

- Resolve [#164](https://github.com/dooboolab/flutter_calendar_carousel/issues/164).

## [1.4.0]

- Resolve [#154](https://github.com/dooboolab/flutter_calendar_carousel/issues/154).

## [1.3.29]

- Resolve [#157](https://github.com/dooboolab/flutter_calendar_carousel/issues/157).

## [1.3.28]

- Allow the use of generic type with Interface [#149](https://github.com/dooboolab/flutter_calendar_carousel/pull/149)
- Added doc to custom weekday builder. Weekday number is now supplied to the builder [#150](https://github.com/dooboolab/flutter_calendar_carousel/pull/150)

## [1.3.27]

- customDayBuilder fix.
- Remove date_utils dep.

## [1.3.26]

- Support custom day container feature [#145](https://github.com/dooboolab/flutter_calendar_carousel/pull/145).

## [1.3.23]

- Support intl >= 0.15.7 < 0.17.0 to inclease `pub` health
- Removed deprecated methods ~~`markedDates`~~, ~~`markedDateColor`~~
- Fixes [#101](https://github.com/dooboolab/flutter_calendar_carousel/issues/101)
- Fixes [#104](https://github.com/dooboolab/flutter_calendar_carousel/issues/104)
- Fixes [#112](https://github.com/dooboolab/flutter_calendar_carousel/issues/112)
- Fixes [#119](https://github.com/dooboolab/flutter_calendar_carousel/issues/119)
- Support long pressed as a feature request[#103](https://github.com/dooboolab/flutter_calendar_carousel/issues/103)
- Support semantic label as a feature request [#139](https://github.com/dooboolab/flutter_calendar_carousel/issues/139)
- Expose `dayCrossAxisAlignment` and `dayMainAxisAlignment` to resolve [#122](https://github.com/dooboolab/flutter_calendar_carousel/issues/122)
- Expose `showIconBehindDayText` to resolve [#131](https://github.com/dooboolab/flutter_calendar_carousel/issues/131)
- Fixes [#94](https://github.com/dooboolab/flutter_calendar_carousel/issues/94)

## [1.3.20]

- Support intl >= 0.15.7

## [1.3.19]

- Improved customizability for weekday containers [#141](https://github.com/dooboolab/flutter_calendar_carousel/pull/141)

## [1.3.18]

- Fix vertical scroll behavior for weekFormat calendar view.
- Reformat code with dartfmt

## [1.3.17]

- Added feature to only show dates from today adding `showOnlyCurrentMonthDate` parameter.

## [1.3.16]

- Added feature for change first day of the week.

## [1.3.15+]

- Ability to disable horizontal scroll to change month with `isScrollable` param.
  - Resolve [#74](https://github.com/dooboolab/flutter_calendar_carousel/issues/74)
- Show events in `week` calendar.
  - Resolve [#66](https://github.com/dooboolab/flutter_calendar_carousel/issues/66)
- Update breaking docs in pub.
- Expose event list to user.
- Remove print.
- Pass first date of week to onCalendarChanged in week view [#88](https://github.com/dooboolab/flutter_calendar_carousel/pull/88)
- Support for passing in custom widgets for next and previous month arrow icons [#95](https://github.com/dooboolab/flutter_calendar_carousel/pull/95)

## [1.3.14]

- Code refactoring [#77](https://github.com/dooboolab/flutter_calendar_carousel/pull/77)
  - Seperate weekday widget

## [1.3.13]

- Code refactoring [#73](https://github.com/dooboolab/flutter_calendar_carousel/pull/73)
  - Seperate header.
  - Add first basic test code.
  - Add composable header widget

## [1.3.12]

- Setting dot icon per event [#71](https://github.com/dooboolab/flutter_calendar_carousel/pull/71)

## [1.3.11]

- Fixed `selectledDayTextStyle` property not being respected [#65](https://github.com/dooboolab/flutter_calendar_carousel/pull/65).

## [1.3.10]

- Add property for static six week format to keep calendar height consistent between months [#62](https://github.com/dooboolab/flutter_calendar_carousel/pull/62).

## [1.3.9]

- Changed priority for `today` higher than `prevMonth` and `nextMonth` and `thisMonth`.

## [1.3.7]

- Ability to choose the weekday format on the constructor [#47](https://github.com/dooboolab/flutter_calendar_carousel/pull/47).

## [1.3.6]

- custom event type added [#49](https://github.com/dooboolab/flutter_calendar_carousel/pull/49).

## [1.3.5]

- headerTitleTouchable and onHeaderTitlePressed props added [#44](https://github.com/dooboolab/flutter_calendar_carousel/pull/44).

## [1.3.4]


test/src/header_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter/material.dart';
import 'package:slidy_calendar/src/calendar_header.dart';


void main() {
  final title = "Test title";
  final margin = const EdgeInsets.symmetric(vertical: 16.0);
  final iconColor = Colors.blueAccent;

  testWidgets('Verify Header Defaults', (WidgetTester tester) async {
    var headerTapped = false;
    var leftPressed = false;
    var rightPressed = false;

    await tester.pumpWidget(wrapped(CalendarHeader(
      headerTitle: title,
      headerMargin: margin,
      showHeader: true,
      showHeaderButtons: true,
      headerIconColor: iconColor,
      onHeaderTitlePressed: () => headerTapped = true,
      onRightButtonPressed: () => rightPressed = true,
      onLeftButtonPressed: () => leftPressed = true,
      isTitleTouchable: true,
    )));

    expect(find.text(title), findsOneWidget);

    await tester.tap(find.byType(FlatButton));

    await tester.pump();

    expect(headerTapped, equals(true));

    await tester.tap(find.widgetWithIcon(IconButton, Icons.chevron_right));

    await tester.pump();

    expect(rightPressed, equals(true));

    await tester.tap(find.widgetWithIcon(IconButton, Icons.chevron_left));

    await tester.pump();

    expect(leftPressed, equals(true));
  });

  testWidgets('Verify No header Renders', (WidgetTester tester) async {
    final noHeaderEmpty = CalendarHeader(
      showHeader: false,
      headerTitle: null,
      onLeftButtonPressed: () {},
      onHeaderTitlePressed: () {},
      onRightButtonPressed: () {},
    );

    await tester.pumpWidget(Container(child: noHeaderEmpty));

    expect(find.byWidget(noHeaderEmpty), findsOneWidget);
  });

  testWidgets('Verify Header Is Not Touchable', (WidgetTester tester) async {
    await tester.pumpWidget(wrapped(CalendarHeader(
      headerTitle: title,
      headerMargin: margin,
      showHeader: true,
      showHeaderButtons: true,
      headerIconColor: iconColor,
      onHeaderTitlePressed: () {},
      onRightButtonPressed: () {},
      onLeftButtonPressed: () {},
      isTitleTouchable: false,
    )));

    // the header FlatButton Should not render
    final touchableHeader = find.byType(FlatButton);

    expect(touchableHeader, findsNothing);
  });

  testWidgets('Verify No Header Buttons', (WidgetTester tester) async {
    await tester.pumpWidget(wrapped(CalendarHeader(
      headerTitle: title,
      headerMargin: margin,
      showHeader: true,
      showHeaderButtons: false,
      headerIconColor: iconColor,
      onHeaderTitlePressed: () {},
      onRightButtonPressed: () {},
      onLeftButtonPressed: () {},
      isTitleTouchable: true,
    )));

    // the header IconButtons Should not render
    final headerButton = find.byType(IconButton);

    expect(headerButton, findsNothing);
  });
}

// header uses Row which requires MaterialApp as an ancestor
Widget wrapped(Widget widget) => MaterialApp(
      home: Container(
        child: Material(child: widget),
      ),
    );


example/test/widget_test.dart
// This is a basic Flutter widget test.
// To perform an interaction with a widget in your test, use the WidgetTester utility that Flutter
// provides. For example, you can send tap and scroll gestures. You can also use WidgetTester to
// find child widgets in the widget tree, read text, and verify that the values of widget properties
// are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:slidy_calendar/slidy_calendar.dart';

void main() {
  DateTime pressedDay;
  testWidgets('Default test for Calendar Carousel',
      (WidgetTester tester) async {
    //  Build our app and trigger a frame.
    await tester.pumpWidget(MaterialApp(
      home: Scaffold(
        body: Container(
          child: SlidyCalendar(
            daysHaveCircularBorder: null,
            weekendTextStyle: TextStyle(
              color: Colors.red,
            ),
            thisMonthDayBorderColor: Colors.grey,
            headerText: 'Custom Header',
            weekFormat: true,
            height: 200.0,
            showIconBehindDayText: true,
            customGridViewPhysics: NeverScrollableScrollPhysics(),
            markedDateShowIcon: true,
            markedDateIconMaxShown: 2,
            selectedDayTextStyle: TextStyle(
              color: Colors.yellow,
            ),
            todayTextStyle: TextStyle(
              color: Colors.blue,
            ),
            markedDateIconBuilder: (event) {
              return event.icon;
            },
            todayButtonColor: Colors.transparent,
            todayBorderColor: Colors.green,
            markedDateMoreShowTotal:
                true, // null for not showing hidden events indicator
            onDayPressed: (date, event) {
              pressedDay = date;
            },
          ),
        ),
      ),
    ));

    expect(find.byType(SlidyCalendar), findsOneWidget);

    expect(pressedDay, isNull);

    await tester.tap(find.text(DateTime.now().day.toString().padLeft(2, '0')));

    await tester.pump();

    expect(pressedDay, isNotNull);
  });
}


test/slidy_calendar_test.dart
// This is a basic Flutter widget test.
// To perform an interaction with a widget in your test, use the WidgetTester utility that Flutter
// provides. For example, you can send tap and scroll gestures. You can also use WidgetTester to
// find child widgets in the widget tree, read text, and verify that the values of widget properties
// are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:slidy_calendar/slidy_calendar.dart';

void main() {
  DateTime pressedDay;
  testWidgets('Default test for Calendar Carousel', (WidgetTester tester) async {
    //  Build our app and trigger a frame.
    await tester.pumpWidget(MaterialApp(
      home: Scaffold(
        body: Container(
          child: SlidyCalendar(
            daysHaveCircularBorder: null,
            weekendTextStyle: TextStyle(
              color: Colors.red,
            ),
            thisMonthDayBorderColor: Colors.grey,
            headerText: 'Custom Header',
            weekFormat: true,
            height: 200.0,
            showIconBehindDayText: true,
            customGridViewPhysics: NeverScrollableScrollPhysics(),
            markedDateShowIcon: true,
            markedDateIconMaxShown: 2,
            selectedDayTextStyle: TextStyle(
              color: Colors.yellow,
            ),
            todayTextStyle: TextStyle(
              color: Colors.blue,
            ),
            markedDateIconBuilder: (event) {
              return event.icon;
            },
            todayButtonColor: Colors.transparent,
            todayBorderColor: Colors.green,
            markedDateMoreShowTotal: true, // null for not showing hidden events indicator
            onDayPressed: (date, event) {
              pressedDay = date;
            },
          ),
        ),
      ),
    ));

    expect(find.byType(SlidyCalendar), findsOneWidget);
  });

  testWidgets(
    'make sure onDayPressed is called when the user tap',
    (WidgetTester tester) async {
      DateTime pressedDay;

      await tester.pumpWidget(
        MaterialApp(
          home: Scaffold(
            body: Container(
              child: SlidyCalendar(
                weekFormat: true,
                height: 200.0,
                onDayPressed: (date, event) {
                  pressedDay = date;
                },
              ),
            ),
          ),
        ),
      );

      expect(find.byType(SlidyCalendar), findsOneWidget);

      expect(pressedDay, isNull);

      await tester.tap(find.text(DateTime.now().day.toString()));

      await tester.pump();

      expect(pressedDay, isNotNull);
    },
  );

  testWidgets(
    'should do nothing when the user tap and onDayPressed is not provided',
    (WidgetTester tester) async {
      await tester.pumpWidget(
        MaterialApp(
          home: Scaffold(
            body: Container(
              child: SlidyCalendar(
                weekFormat: true,
                height: 200.0,
              ),
            ),
          ),
        ),
      );

      expect(find.byType(SlidyCalendar), findsOneWidget);

      await tester.tap(find.text(DateTime.now().day.toString()));
      await tester.pump();
    },
  );

  testWidgets(
    'make sure onDayLongPressed is called when the user press and hold',
    (WidgetTester tester) async {
      DateTime longPressedDay;

      await tester.pumpWidget(
        MaterialApp(
          home: Scaffold(
            body: Container(
              child: SlidyCalendar(
                weekFormat: true,
                height: 200.0,
                onDayLongPressed: (date) {
                  longPressedDay = date;
                },
              ),
            ),
          ),
        ),
      );

      expect(find.byType(SlidyCalendar), findsOneWidget);

      expect(longPressedDay, isNull);

      await tester.longPress(find.text(DateTime.now().day.toString()));
      await tester.pump();

      expect(longPressedDay, isNotNull);
    },
  );

  testWidgets(
    'should do nothing when the user press and hold and onDayLongPressed is not provided',
    (WidgetTester tester) async {
      await tester.pumpWidget(
        MaterialApp(
          home: Scaffold(
            body: Container(
              child: SlidyCalendar(
                weekFormat: true,
                height: 200.0,
              ),
            ),
          ),
        ),
      );

      expect(find.byType(SlidyCalendar), findsOneWidget);

      await tester.longPress(find.text(DateTime.now().day.toString()));
      await tester.pump();
    },
  );
}


lib/classes/event.dart
import 'package:flutter/material.dart';

class Event implements EventInterface {
  final DateTime date;
  final String title;
  final Widget icon;
  final Widget dot;

  Event({
    this.date,
    this.title,
    this.icon,
    this.dot,
  }) : assert(date != null);

  @override
  bool operator ==(dynamic other) {
    return date == other.date && title == other.title && icon == other.icon && dot == other.dot;
  }

  @override
  int get hashCode => hashValues(date, title, icon);

  @override
  DateTime getDate() {
    return date;
  }

  @override
  Widget getDot() {
    return dot;
  }

  @override
  Widget getIcon() {
    return icon;
  }

  @override
  String getTitle() {
    return title;
  }
}

abstract class EventInterface {
  DateTime getDate();
  String getTitle();
  Widget getIcon();
  Widget getDot();
}


example/lib/main.dart
import 'package:flutter/material.dart';

import 'package:intl/intl.dart' show DateFormat;

import 'package:slidy_calendar/slidy_calendar.dart';

import 'package:slidy_calendar/classes/event.dart';
import 'package:slidy_calendar/classes/event_list.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'dooboolab flutter calendar',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or press Run > Flutter Hot Reload in IntelliJ). Notice that the
        // counter didn't reset back to zero; the application is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Calendar Carousel Example'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  DateTime _currentDate = DateTime(2019, 2, 3);
  DateTime _currentDate2 = DateTime(2019, 2, 3);
  String _currentMonth = DateFormat.yMMM().format(DateTime(2019, 2, 3));
  DateTime _targetDateTime = DateTime(2019, 2, 3);
//  List<DateTime> _markedDate = [DateTime(2018, 9, 20), DateTime(2018, 10, 11)];
  static final Widget _eventIcon = Container(
    decoration: BoxDecoration(
        color: Colors.white,
        borderRadius: BorderRadius.all(Radius.circular(1000)),
        border: Border.all(color: Colors.blue, width: 2.0)),
    child: Icon(
      Icons.person,
      color: Colors.amber,
    ),
  );

  final EventList<Event> _markedDateMap = EventList<Event>(
    events: {
      DateTime(2019, 2, 10): [
        Event(
          date: DateTime(2019, 2, 10),
          title: 'Event 1',
          icon: _eventIcon,
          dot: Container(
            margin: EdgeInsets.symmetric(horizontal: 1.0),
            color: Colors.red,
            height: 5.0,
            width: 5.0,
          ),
        ),
        Event(
          date: DateTime(2019, 2, 10),
          title: 'Event 2',
          icon: _eventIcon,
        ),
        Event(
          date: DateTime(2019, 2, 10),
          title: 'Event 3',
          icon: _eventIcon,
        ),
      ],
    },
  );

  SlidyCalendar _calendarCarousel, _calendarCarouselNoHeader;

  @override
  void initState() {
    /// Add more events to _markedDateMap EventList
    _markedDateMap.add(
        DateTime(2019, 2, 25),
        Event(
          date: DateTime(2019, 2, 25),
          title: 'Event 5',
          icon: _eventIcon,
        ));

    _markedDateMap.add(
        DateTime(2019, 2, 10),
        Event(
          date: DateTime(2019, 2, 10),
          title: 'Event 4',
          icon: _eventIcon,
        ));

    _markedDateMap.addAll(DateTime(2019, 2, 11), [
      Event(
        date: DateTime(2019, 2, 11),
        title: 'Event 1',
        icon: _eventIcon,
      ),
      Event(
        date: DateTime(2019, 2, 11),
        title: 'Event 2',
        icon: _eventIcon,
      ),
      Event(
        date: DateTime(2019, 2, 11),
        title: 'Event 3',
        icon: _eventIcon,
      ),
    ]);
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    /// Example with custom icon
    _calendarCarousel = SlidyCalendar<Event>(
      onDayPressed: (date, events) {
        setState(() => _currentDate = date);
        events.forEach((event) => print(event.title));
      },
      weekendTextStyle: TextStyle(
        color: Colors.red,
      ),
      thisMonthDayBorderColor: Colors.grey,
//          weekDays: null, /// for pass null when you do not want to render weekDays
      headerText: 'Custom Header',
      weekFormat: true,
      markedDatesMap: _markedDateMap,
      height: 200.0,
      selectedDateTime: _currentDate2,
      showIconBehindDayText: true,
//          daysHaveCircularBorder: false, /// null for not rendering any border, true for circular border, false for rectangular border
      customGridViewPhysics: NeverScrollableScrollPhysics(),
      markedDateShowIcon: true,
      markedDateIconMaxShown: 2,
      selectedDayTextStyle: TextStyle(
        color: Colors.yellow,
      ),
      todayTextStyle: TextStyle(
        color: Colors.blue,
      ),
      markedDateIconBuilder: (event) {
        return event.icon;
      },
      minSelectedDate: _currentDate.subtract(Duration(days: 360)),
      maxSelectedDate: _currentDate.add(Duration(days: 360)),
      todayButtonColor: Colors.transparent,
      todayBorderColor: Colors.green,
      markedDateMoreShowTotal: true, // null for not showing hidden events indicator
//          markedDateIconMargin: 9,
//          markedDateIconOffset: 3,
    );

    /// Example Calendar Carousel without header and custom prev & next button
    _calendarCarouselNoHeader = SlidyCalendar<Event>(
      todayBorderColor: Colors.green,
      onDayPressed: (date, events) {
        setState(() => _currentDate2 = date);
        events.forEach((event) => print(event.title));
      },
      daysHaveCircularBorder: true,
      showOnlyCurrentMonthDate: false,
      weekendTextStyle: TextStyle(
        color: Colors.red,
      ),
      thisMonthDayBorderColor: Colors.grey,
      weekFormat: false,
//      firstDayOfWeek: 4,
      markedDatesMap: _markedDateMap,
      height: 420.0,
      selectedDateTime: _currentDate2,
      targetDateTime: _targetDateTime,
      customGridViewPhysics: NeverScrollableScrollPhysics(),
      markedDateCustomShapeBorder: CircleBorder(side: BorderSide(color: Colors.yellow)),
      markedDateCustomTextStyle: TextStyle(
        fontSize: 18,
        color: Colors.blue,
      ),
      showHeader: false,
      todayTextStyle: TextStyle(
        color: Colors.blue,
      ),
      // markedDateShowIcon: true,
      // markedDateIconMaxShown: 2,
      // markedDateIconBuilder: (event) {
      //   return event.icon;
      // },
      // markedDateMoreShowTotal:
      //     true,
      todayButtonColor: Colors.yellow,
      selectedDayTextStyle: TextStyle(
        color: Colors.yellow,
      ),
      minSelectedDate: _currentDate.subtract(Duration(days: 360)),
      maxSelectedDate: _currentDate.add(Duration(days: 360)),
      prevDaysTextStyle: TextStyle(
        fontSize: 16,
        color: Colors.pinkAccent,
      ),
      inactiveDaysTextStyle: TextStyle(
        color: Colors.tealAccent,
        fontSize: 16,
      ),
      onCalendarChanged: ( date) {
        setState(() {
          _targetDateTime = date;
          _currentMonth = DateFormat.yMMM().format(_targetDateTime);
        });
      },
      onDayLongPressed: ( date) {
        print('long pressed date $date');
      },
    );

    return Scaffold(
        appBar: AppBar(
          title: Text(widget.title),
        ),
        body: SingleChildScrollView(
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            mainAxisAlignment: MainAxisAlignment.start,
            children: <Widget>[
              //custom icon
              Container(
                margin: EdgeInsets.symmetric(horizontal: 16.0),
                child: _calendarCarousel,
              ), // This trailing comma makes auto-formatting nicer for build methods.
              //custom icon without header
              Container(
                margin: EdgeInsets.only(
                  top: 30.0,
                  bottom: 16.0,
                  left: 16.0,
                  right: 16.0,
                ),
                child: Row(
                  children: <Widget>[
                    Expanded(
                        child: Text(
                      _currentMonth,
                      style: TextStyle(
                        fontWeight: FontWeight.bold,
                        fontSize: 24.0,
                      ),
                    )),
                    FlatButton(
                      child: Text('PREV'),
                      onPressed: () {
                        setState(() {
                          _targetDateTime = DateTime(_targetDateTime.year, _targetDateTime.month - 1);
                          _currentMonth = DateFormat.yMMM().format(_targetDateTime);
                        });
                      },
                    ),
                    FlatButton(
                      child: Text('NEXT'),
                      onPressed: () {
                        setState(() {
                          _targetDateTime = DateTime(_targetDateTime.year, _targetDateTime.month + 1);
                          _currentMonth = DateFormat.yMMM().format(_targetDateTime);
                        });
                      },
                    )
                  ],
                ),
              ),
              Container(
                margin: EdgeInsets.symmetric(horizontal: 16.0),
                child: _calendarCarouselNoHeader,
              ), //
            ],
          ),
        ));
  }
}
